File: semana_4_5_patrones_creacionales/desarrollo/desarrollo1_singleton_logger.cpp
/**
 * DESARROLLO 1: Singleton en C++ - Sistema de Logging
 * (EJERCICIO NUEVO - NO es el ejemplo del PDF)
 * 
 * Sistema simple de logging con Singleton
 */

#include <iostream>
#include <fstream>
#include <string>
using namespace std;

class Logger {
private:
    static Logger* instance;
    ofstream logFile;
    
    Logger() {
        logFile.open("aplicacion.log", ios::app);
    }

public:
    static Logger* getInstance() {
        if (instance == nullptr) {
            instance = new Logger();
        }
        return instance;
    }
    
    void log(const string& mensaje) {
        if (logFile.is_open()) {
            logFile << mensaje << endl;
            cout << "[LOG] " << mensaje << endl;
        }
    }
    
    ~Logger() {
        if (logFile.is_open()) {
            logFile.close();
        }
    }
};

Logger* Logger::instance = nullptr;

int main() {
    cout << "=== DESARROLLO 1: SINGLETON - LOGGER ===" << endl;
    
    Logger* logger1 = Logger::getInstance();
    Logger* logger2 = Logger::getInstance();
    
    cout << "Son la misma instancia? " << (logger1 == logger2 ? "SI" : "NO") << endl;
    
    logger1->log("Aplicacion iniciada");
    logger1->log("Usuario admin logueado");
    logger2->log("Consulta ejecutada");
    
    return 0;
}

/**
 * COMPILAR: g++ -std=c++11 desarrollo1_singleton_logger.cpp -o logger
 * EJECUTAR: ./logger
 */

File: semana_4_5_patrones_creacionales/desarrollo/desarrollo1_singleton_logger.py
"""
DESARROLLO 1: Singleton en Python - Sistema de Logging
(VERSION CRUZADA del desarrollo C++)
"""

class Logger:
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if self._initialized:
            return
        self._initialized = True
        self.log_file = open("aplicacion.log", "a")
    
    def log(self, mensaje):
        self.log_file.write(mensaje + "\n")
        print(f"[LOG] {mensaje}")
    
    def __del__(self):
        if hasattr(self, 'log_file'):
            self.log_file.close()


def main():
    print("=== DESARROLLO 1: SINGLETON - LOGGER (Python) ===")
    
    logger1 = Logger()
    logger2 = Logger()
    
    print(f"Son la misma instancia? {logger1 is logger2}")
    
    logger1.log("Aplicacion iniciada")
    logger1.log("Usuario admin logueado")
    logger2.log("Consulta ejecutada")


if __name__ == "__main__":
    main()

"""
EJECUTAR: python3 desarrollo1_singleton_logger.py
"""
